doc/spec/json.md

```diff
@@ -0,0 +1,87 @@
+# Docker Distribution JSON Canonicalization
+
+## Introduction
+
+To provide consistent content hashing of JSON object throughout Docker
+Distribution APIs, a canonical JSON format has been defined. The
+canonicalization is balance between conistency and readability. The format
+also provides consistent output for effective caching.
+
+## Rules
+
+1. Object keys shall always appear in lexically sorted order.
+2. The colon (":") shall always be followed by a single space.
+3. An indent of three spaces ("   ") shall be used.
+4. The number of indents used shall be proportional to the current depth of
+   nesting, where the depth increments after the start of an array or object
+   and decrements afterwards.
+5. The comma (",") shall always be followed by a newline and the number of
+   indents corresponding to the current depth.
+6. Empty objects ("{}") and arrays ("[]") shall not have any internal spacing.
+7. No "trailing commas" are allowed in object or array definitions.
+
+## Examples
+
+The following is a simple example of a canonicalized JSON string:
+
+```json
+{
+   "asdf": 1,
+   "qwer": [],
+   "zxcv": [
+      {},
+      true,
+      1000000000,
+      "tyui"
+   ]
+}
+```
+
+## Reference
+
+### Other Canonicalizations
+
+The OLPC project specifies [Canonical
+JSON](http://wiki.laptop.org/go/Canonical_JSON). While this is used in TUF,
+which may be used with other distribution-related protocols, this alternative
+format has been proposed for readability. Specifications complying with either
+this specification or an alternative should explicitly call out the
+canonicalization format.
+
+### Go
+
+In Go, the canonical JSON format spacing can be obtained by calling the
+[`json.Indent`](http://golang.org/src/encoding/json/indent.go?s=1918:1989#L65)
+with the following arguments:
+
+```go
+incoming := getBytes()
+var canonical bytes.Buffer
+if err := json.Indent(&canonical, incoming, "", "   "); err != nil {
+	// ... handle error
+}
+```
+
+By default, [`encoding/json`](http://golang.org/pkg/encoding/json/) will
+generate objects with sorted keys. Simply using `json.MarshalIndent` with the
```

Comment:
Seems likely to cause confusion since there is no way to determine by looking at map whether keys are in a "particular" order or lexical order. The problem being that a JSON unmarshall + marshall may not produce the same result unless explicitly decoded into such structure. If this is allowed it seems pointless to even define a canonical JSON, and just leave it as a suggested format.